packages/tx/src/EIP2930Transaction.ts

```diff
@@ -0,0 +1,391 @@
+      const address = <Buffer>accessListItem[0]
+      const storageSlots = <Buffer[]>accessListItem[1]
+      if ((<any>accessListItem)[2] !== undefined) {
+        throw new Error(
+          'Access list item cannot have 3 elements. It can only have an address, and an array of storage slots.'
+        )
+      }
+      if (address.length != 20) {
+        throw new Error('Invalid EIP-2930 transaction: address length should be 20 bytes')
+      }
+      for (let storageSlot = 0; storageSlot < storageSlots.length; storageSlot++) {
+        if (storageSlots[storageSlot].length != 32) {
+          throw new Error('Invalid EIP-2930 transaction: storage slot length should be 32 bytes')
+        }
+      }
+    }
+
+    const freeze = opts?.freeze ?? true
+    if (freeze) {
+      Object.freeze(this)
+    }
+  }
+
+  getMessageToSign() {
+    const base = [
+      bnToRlp(this.chainId),
+      bnToRlp(this.nonce),
+      bnToRlp(this.gasPrice),
+      bnToRlp(this.gasLimit),
+      this.to !== undefined ? this.to.buf : Buffer.from([]),
+      bnToRlp(this.value),
+      this.data,
+      this.accessList,
+    ]
+    return keccak256(Buffer.concat([Buffer.from('01', 'hex'), rlp.encode(base)]))
+  }
+
+  /**
+   * The amount of gas paid for the data in this tx
+   */
+  getDataFee(): BN {
+    const cost = super.getDataFee()
+    const accessListStorageKeyCost = this.common.param('gasPrices', 'accessListStorageKeyCost')
+    const accessListAddressCost = this.common.param('gasPrices', 'accessListAddressCost')
+
+    let slots = 0
+    for (let index = 0; index < this.accessList.length; index++) {
+      const item = this.accessList[index]
+      const storageSlots = item[1]
+      slots += storageSlots.length
+    }
+
+    const addresses = this.accessList.length
+    cost.iaddn(addresses * accessListAddressCost + slots * accessListStorageKeyCost)
+    return cost
+  }
+
+  /**
+   * Returns a Buffer Array of the raw Buffers of this transaction, in order.
+   * @param asList - By default, this method returns a concatenated Buffer
+   *                 If this is not desired, then set this to `true`, to get a Buffer array.
+   */
+  raw(asList = false): Buffer[] | Buffer {
+    const base = <Buffer[]>[
+      bnToRlp(this.chainId),
+      bnToRlp(this.nonce),
+      bnToRlp(this.gasPrice),
+      bnToRlp(this.gasLimit),
+      this.to !== undefined ? this.to.buf : Buffer.from([]),
+      bnToRlp(this.value),
+      this.data,
+      this.accessList,
+      this.v !== undefined ? bnToRlp(this.v) : Buffer.from([]),
+      this.r !== undefined ? bnToRlp(this.r) : Buffer.from([]),
+      this.s !== undefined ? bnToRlp(this.s) : Buffer.from([]),
+    ]
+    if (!asList) {
+      return Buffer.concat([Buffer.from('01', 'hex'), rlp.encode(base)])
+    } else {
+      return base
+    }
+  }
+
+  /**
+   * Returns the encoding of the transaction. For typed transaction, this is the raw Buffer.
+   * In LegacyTransaction, this is a Buffer array.
```

Comment:
why different between Legacy and Typed?

Reply:
In a block, a legacy transaction is encoded as a buffer array, while a typed transaction is a raw buffer. The typed transaction starts with a byte which represents the transaction type, and then we concatenate the transaction payload, which is usually the rlp encoding of the data.